src/shared/sleep-config.c

```diff
@@ -102,6 +102,120 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+
+        log_debug("Current battery charge percentage: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+static int get_battery_discharge_rate_filepath(char **ret_path) {
+        sd_id128_t machine_id;
+        int r;
+
+        assert(ret_path);
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get machine ID: %m");
+
+        *ret_path = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_percentage_rate_per_hour");
+        if (!*ret_path)
+                return log_oom_debug();
+        return 0;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *filepath = NULL;
+        char *discharge_rate;
+        int stored_discharge_rate, r;
+
+        r = get_battery_discharge_rate_filepath(&filepath);
+        if (r < 0)
+               return r;
+
+        r = read_one_line_file(filepath, &discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read discharge rate from %s: %m", filepath);
+
+        r = safe_atoi(discharge_rate, &stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", filepath);
+
+        if (stored_discharge_rate > 0 && stored_discharge_rate < 200)
+               return stored_discharge_rate;
+        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid battery discharge percentage rate per hour: %m");
+}
+
+int put_battery_discharge_rate(int estimated_battery_discharge_rate) {
+        _cleanup_free_ char *filepath = NULL;
+        int r;
+
+        r = get_battery_discharge_rate_filepath(&filepath);
+        if (r < 0)
+               return r;
+
+        r = write_string_filef(filepath, WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_MKDIR_0755, "%d", estimated_battery_discharge_rate);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to create %s: %m", filepath);
+
+        return 0;
+}
+
+int battery_is_alarming(void) {
+        _cleanup_free_ char *battery_energy_now_string = NULL, *battery_alarm_string = NULL;
+        int battery_energy_now, battery_alarm, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/energy_now", &battery_energy_now_string);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/energy_now: %m");
+
+        r = safe_atoi(battery_energy_now_string, &battery_energy_now);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery present energy: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/alarm", &battery_alarm_string);
```

Comment:
don't we have to set up the alarm first? i mean, i assume by default laptops won't wake up from suspend when the alarm is configured, we must enable that first, no?